engine/consensus/matching/engine.go

```diff
@@ -0,0 +1,187 @@
+// algorithm has started.
+func (e *Engine) Ready() <-chan struct{} {
+	e.unit.Launch(e.loop)
+	return e.unit.Ready()
+}
+
+// Done returns a done channel that is closed once the engine has fully stopped.
+// For the consensus engine, we wait for hotstuff to finish.
+func (e *Engine) Done() <-chan struct{} {
+	return e.unit.Done()
+}
+
+// SubmitLocal submits an event originating on the local node.
+func (e *Engine) SubmitLocal(event interface{}) {
+	e.Submit(e.me.NodeID(), event)
+}
+
+// Submit submits the given event from the node with the given origin ID
+// for processing in a non-blocking manner. It returns instantly and logs
+// a potential processing error internally when done.
+func (e *Engine) Submit(originID flow.Identifier, event interface{}) {
+	err := e.Process(originID, event)
+	if err != nil {
+		e.log.Fatal().Err(err).Msg("internal error processing event")
+	}
+}
+
+// ProcessLocal processes an event originating on the local node.
+func (e *Engine) ProcessLocal(event interface{}) error {
+	return e.Process(e.me.NodeID(), event)
+}
+
+// Process processes the given event from the node with the given origin ID in
+// a blocking manner. It returns the potential processing error when done.
+func (e *Engine) Process(originID flow.Identifier, event interface{}) error {
+	receipt, ok := event.(*flow.ExecutionReceipt)
+	if !ok {
+		return fmt.Errorf("input message of incompatible type: %T, origin: %x", event, originID[:])
+	}
+	e.metrics.MessageReceived(metrics.EngineSealing, metrics.MessageExecutionReceipt)
+	e.pendingReceipts.Put(&engine.Message{OriginID: originID, Payload: receipt})
+	e.notifier.Notify()
+	return nil
+}
+
+// HandleReceipt ingests receipts from the Requester module.
+func (e *Engine) HandleReceipt(originID flow.Identifier, receipt flow.Entity) {
+	e.log.Debug().Msg("received receipt from requester engine")
+	e.metrics.MessageReceived(metrics.EngineSealing, metrics.MessageExecutionReceipt)
+	e.pendingReceipts.Put(&engine.Message{OriginID: originID, Payload: receipt})
+	e.notifier.Notify()
+}
+
+// OnFinalizedBlock implements the `OnFinalizedBlock` callback from the `hotstuff.FinalizationConsumer`
+// CAUTION: the input to this callback is treated as trusted; precautions should be taken that messages
+// from external nodes cannot be considered as inputs to this function
+func (e *Engine) OnFinalizedBlock(finalizedBlockID flow.Identifier) {
+	e.pendingFinalizationEvents.Push(finalizedBlockID)
+	e.notifier.Notify()
+}
+
+func (e *Engine) loop() {
+	c := e.notifier.Channel()
+	for {
+		select {
+		case <-e.unit.Quit():
+			return
+		case <-c:
+			err := e.processAvailableEvents()
+			if err != nil {
+				e.log.Fatal().Err(err).Msg("internal error processing queued message")
+			}
+		}
+	}
+}
+
+// processAvailableEvents processes _all_ available events (untrusted messages
+// from other nodes as well as internally trusted
+func (e *Engine) processAvailableEvents() error {
+	for {
```

Comment:
A routine could potentially spend quite some time in this loop. Therefore, I think we should include a termination statement in case the Engine is shut down: 

```suggestion
	for {
		select {
		case <-e.unit.Quit():
			return nil
		default:
		}

```